### ISSUE TITLE
OnLoadEvent not being fired

### ISSUE THREAD

**Issue by Daniel-Houston** [mkenney/go-chrome#135]
### What version of `go-chrome` are you using (tag, hash, etc.)?
v1.0.0-rc7

### What behavior do you expect? What are you trying to accomplish?
I am trying to print a PDF. I expect it to be printed when the 'page load' event is fired.

### What behavior are you experiencing instead?
The chrome browser loads and the program hangs until the timeout.

Sample output:

```
➜  tester go run main.go
 info Starting process
   ⇢  flags=--addr=localhost --disable-extensions --disable-gpu --headless --hide-scrollbars --no-first-run --no-sandbox --port=9222 --remote-debugging-address=0.0.0.0 --remote-debugging-port=9222 --user-data-dir=/var/folders/bz/84vb3m3d4sd24h07n24_kzq1gx0ldc/T/ path="/Applications/Google Chrome.app/Contents/MacOS/Google Chrome"
   ⇢  chrome.chromium.go:238 github.com/mkenney/go-chrome/tot.(*Chrome).Launch
   ⇢  2019-06-19T16:06:52.996-06:00
 info New socket connection listening
   ⇢  socketID=1 url="ws://localhost:9222/devtools/page/49291603D1EBC96427F879A4F1581100"
   ⇢  socket.socketer.go:74 github.com/mkenney/go-chrome/tot/socket.New
   ⇢  2019-06-19T16:06:54.007-06:00
 info Websocket connection established
   ⇢  status="101 WebSocket Protocol Handshake" url="ws://localhost:9222/devtools/page/49291603D1EBC96427F879A4F1581100"
   ⇢  socket.web_socketer.go:37 github.com/mkenney/go-chrome/tot/socket.NewWebsocket
   ⇢  2019-06-19T16:06:54.010-06:00
```

### How can this behavior be reproduced?
Run this code with `go run main.go`:

```
package main

import (
	"encoding/base64"
	"errors"
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"os/exec"
	"time"

	chrome "github.com/mkenney/go-chrome/tot"
	"github.com/mkenney/go-chrome/tot/page"
)

func main() {

	chromeApp, err := Chrome()
	if err != nil {
		panic(err)
	}

	browser := chrome.New(
		&chrome.Flags{
			"addr":                     "localhost",
			"disable-extensions":       nil,
			"disable-gpu":              nil,
			"headless":                 nil,
			"hide-scrollbars":          nil,
			"no-first-run":             nil,
			"no-sandbox":               nil,
			"port":                     9222,
			"remote-debugging-address": "0.0.0.0",
			"remote-debugging-port":    9222,
		},
		chromeApp, "/tmp", "/dev/null", "/dev/null",
	)

	if err := browser.Launch(); err != nil {
		panic(err)
	}

	websiteURL := "https://www.google.com"
	tab, err := browser.NewTab(websiteURL)
	if err != nil {
		panic(err)
	}

	if enableResult := <-tab.Page().Enable(); nil != enableResult.Err {
		panic(err)
	}

	results := make(chan *page.PrintToPDFResult)

	tab.Page().OnLoadEventFired(
		func(event *page.LoadEventFiredEvent) {
			log.Printf("loaded the thing")

			pdfResult := <-tab.Page().PrintToPDF(&page.PrintToPDFParams{})
			results <- pdfResult
		},
	)

	timeout := time.After(time.Duration(100) * time.Second)

	select {
	case <-timeout:
		panic("Timeout exceeded")
	case result := <-results:
		if result.Err != nil {
			panic(result.Err)
		}

		data, err := base64.StdEncoding.DecodeString(result.Data)
		if err != nil {
			panic(err)
		}

		err = ioutil.WriteFile(buildTemporaryFilename("c02bfcab-2854-448e-8ce0-a4fe3d8ad7d6", "pdf"), data, 0644)
		if err != nil {
			panic(err)
		}
	}
}

func buildTemporaryFilename(id, ext string) string {
	return fmt.Sprintf("%s%c%s.%s", os.TempDir(), os.PathSeparator, id, ext)
}

// Chrome determines that path to Chrome on the system
func Chrome() (string, error) {
	chromeBinaries := []string{
		"google-chrome-stable",
		"chromium-browser",
		"/Applications/Google Chrome.app/Contents/MacOS/Google Chrome",
	}

	for _, chrome := range chromeBinaries {
		path, err := exec.LookPath(chrome)
		if err == nil {
			return path, nil
		}
	}

	return "", errors.New("chrome was not found in your path")
}

```

**Comment by mkenney** [comment]
Hi @Daniel-Houston,

It does return fast, but it looks to me like your script works as intended. I noticed that it creates some zombie processes so I added a `Close()` call and changed the file path for testing.

```diff
*** main.go     2019-06-20 18:53:45.000000000 -0600
--- test.go     2019-06-20 19:10:48.000000000 -0600
***************
*** 37,40 ****
--- 37,41 ----
                chromeApp, "/tmp", "/dev/null", "/dev/null",
        )
+       defer browser.Close()

        if err := browser.Launch(); err != nil {
***************
*** 44,47 ****
--- 45,49 ----
        websiteURL := "https://www.google.com"
        tab, err := browser.NewTab(websiteURL)
+       defer tab.Close()
        if err != nil {
                panic(err)
***************
*** 78,82 ****
                }

!               err = ioutil.WriteFile(buildTemporaryFilename("c02bfcab-2854-448e-8ce0-a4fe3d8ad7d6", "pdf"), data, 0644)
                if err != nil {
                        panic(err)
--- 80,84 ----
                }

!               err = ioutil.WriteFile("/tmp/tst.pdf", data, 0644)
                if err != nil {
                        panic(err)
```

Adding the `Close()` calls uncovered that the socket listeners don't shutdown gracefully and throw some errors, I will take a look at that.

But, it created the pdf I expected: [tst.pdf](https://github.com/mkenney/go-chrome/files/3312612/tst.pdf)

**Comment by mkenney** [comment]
log output:
```
→ go run test.go
 info 2019-06-20T19:17:38.657-06:00 Starting process
   ⇢  flags={"addr":"localhost","disable-extensions":null,"disable-gpu":null,"headless":null,"hide-scrollbars":null,"no-first-run":null,"no-sandbox":null,"port":9222,"remote-debugging-address":"0.0.0.0","remote-debugging-port":9222,"user-data-dir":"/var/folders/0x/xys3g4w97hl8_kkgx3kh2zw00000gp/T/"} path="/Applications/Google Chrome.app/Contents/MacOS/Google Chrome"
   ⇢  chrome.chromium.go:238 github.com/mkenney/tmp/vendor/github.com/mkenney/go-chrome/tot.(*Chrome).Launch
 info 2019-06-20T19:17:39.678-06:00 New socket connection listening
   ⇢  socketID=1 url="ws://localhost:9222/devtools/page/B45B743DCFE4739A319B81778837738F"
   ⇢  socket.socketer.go:74 github.com/mkenney/tmp/vendor/github.com/mkenney/go-chrome/tot/socket.New
 info 2019-06-20T19:17:39.683-06:00 Websocket connection established
   ⇢  status="101 WebSocket Protocol Handshake" url="ws://localhost:9222/devtools/page/B45B743DCFE4739A319B81778837738F"
   ⇢  socket.web_socketer.go:37 github.com/mkenney/tmp/vendor/github.com/mkenney/go-chrome/tot/socket.NewWebsocket
 info 2019-06-20T19:17:40.197-06:00 Executing handler
   ⇢  event="Page.loadEventFired" handler#=0 socketID=1
   ⇢  socket.socketer.go:218 github.com/mkenney/tmp/vendor/github.com/mkenney/go-chrome/tot/socket.(*Socket).handleEvent
2019/06/20 19:17:40 loaded the thing
error 2019-06-20T19:17:41.384-06:00 5001: An unknown error occurred
   ⇢  socketID=1
   ⇢  socket.socketer.go:302 github.com/mkenney/tmp/vendor/github.com/mkenney/go-chrome/tot/socket.(*Socket).listen
error 2019-06-20T19:17:41.385-06:00 nil response from socket
   ⇢  socketID=1
   ⇢  socket.socketer.go:309 github.com/mkenney/tmp/vendor/github.com/mkenney/go-chrome/tot/socket.(*Socket).listen
error 2019-06-20T19:17:41.385-06:00 Unknown response from web socket
   ⇢  data="{\\\"error\\\":null,\\\"id\\\":0,\\\"method\\\":\\\"\\\",\\\"params\\\":null,\\\"result\\\":null}" method="" responseID=0 socketID=1
   ⇢  socket.socketer.go:325 github.com/mkenney/tmp/vendor/github.com/mkenney/go-chrome/tot/socket.(*Socket).listen
 info 2019-06-20T19:17:41.386-06:00 Socket shutting down
   ⇢  socketID=1 url="ws://localhost:9222/devtools/page/B45B743DCFE4739A319B81778837738F"
   ⇢  socket.socketer.go:337 github.com/mkenney/tmp/vendor/github.com/mkenney/go-chrome/tot/socket.(*Socket).listen
 info 2019-06-20T19:17:41.398-06:00 Chromium exited
   ⇢  signal="signal: interrupt"
   ⇢  chrome.chromium.go:132 github.com/mkenney/tmp/vendor/github.com/mkenney/go-chrome/tot.(*Chrome).Close
```

**Comment by Daniel-Houston** [comment]
Interesting that it works for you. When I run it on my computer, I get the first three info messages, ending in 
```
info Websocket connection established
   ⇢  status="101 WebSocket Protocol Handshake" url="ws://localhost:9222/devtools/page/465162C4B03BF774D1A696D59D3C930D"
   ⇢  socket.web_socketer.go:37 github.com/mkenney/go-chrome/tot/socket.NewWebsocket
   ⇢  2019-06-21T09:52:21.872-06:00
```

But I don't get the message saying 
```
 info 2019-06-20T19:17:40.197-06:00 Executing handler
   ⇢  event="Page.loadEventFired" handler#=0 socketID=1
```
Do you have any idea what would be causing the event not to fire on my computer? Is there anything I can do to debug it?

**Comment by mkenney** [comment]
You could try starting Chrome in dev mode from the command line, something like 
```
/Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome --addr=localhost --port=9222 --remote-debugging-port=9222 --remote-debugging-address=0.0.0.0 --disable-extensions --disable-gpu --headless --hide-scrollbars --no-first-run --no-sandbox"
```

and then try your script and see if that makes any difference. Dev mode isn't the default, I'm not sure why it's on for me. I've messed with my Chrome settings a lot over the years though so who knows 🤷‍♂ 

Alternatively, if you use [Docker](https://docs.docker.com/docker-for-mac/install/) there's a [`docker-comopse.yml`](https://github.com/mkenney/go-chrome/blob/master/docker-compose.yml) that will start a headless chrome service that you can access at `localhost:9222`. It might complain about the port if you have any zombie background chrome processes lurking around though, you can just use a different port in the `docker-compose.yml` file.

```
docker-compose up -d chrome
```

Let me know if that helps at all.

**Comment by mkenney** [comment]
The graceful socket shutdown updates are included in `v1.0.0-rc8`.
